Library/Homebrew/service.rb

```diff
@@ -497,5 +500,93 @@ def to_systemd_timer
 
       timer + options.join("\n")
     end
+
+    # Only evaluate the service block once.
+    sig { void }
+    def eval_service_block
+      return if @eval_service_block
+
+      instance_eval(&@service_block)
+      @eval_service_block = true
+    end
+
+    # Recursively prepare the service hash for inclusion in the formula API JSON.
+    # - Replace each Symbol with a String prefixed by ":"
+    # - Replace the local HOMEBREW_PREFIX with "$HOMEBREW_PREFIX"
+    # - Replace the local home directory with "$HOME"
+    def serialize(elem = to_h)
```

Comment:
Can this get a type signature?

Reply:
We could always slap on something generic but I'm not sure how useful a specific type signature would be here. The only other option would be refactoring this into smaller functions.

@dduugg How would you approach adding typing to a recursive function like this?

Reply:
If I'm reading the method documentation correctly, the best you can probably do is:
```ruby
sig { params(elem: T.untyped).returns(T::Hash[String, T.untyped]) }
```
Trying to type an undefined shape generally isn't worth the effort. `JSON.dump` is probably analogous here, and sorbet just leaves it completely [untyped](https://github.com/sorbet/sorbet/blob/60fb704/rbi/stdlib/json.rbi#L130-L138).

Reply:
Oh, another pattern I've seen involves typing the input, something like:
```ruby
sig { params(elem: T::Hash[Symbol, T.untyped]).returns(T::Hash[String, T.untyped]) }
```
and then have `serialize` call a private `_serialize` method that handles the recursive calls. (Ofc, if you know more about the shape of the input/output you could flesh out the type annotations accordingly.)

Reply:
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/bc85857e1341eef2e105cbacd3e5f1e4eaaf47a0